### DISCUSSION TITLE
GLSP-client and Sprotty dependency

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#611] [Q&A] [answered]
I have a question about the versions of GLSP-Client and Sprotty.

In my project I am based on the dependency

     @eclipse-glsp/client": "0.9.0"

as far as I can see this binds the sprotty dependency 

    sprotty 0.10.0

In the moment I am developing  a lot about Routing and `AnchorComputers` to optimize the behaviour of my BPMN core.

I have seen that there were some changes in sprotty in version 0.11.0 and 0.11.1 which also affects Routing. Now I ask myself if sprotty version 0.11.1 would show different behaviour in my project and if I am currently working on problems which may have been already solved...?

Is it possible that I upgrade sprotty to version 0.11.1 in my project? I guess this is not compatible with eclipse-glsp/client 0.9.0 ?

**Comment by tortmayr** [comment, marked as answer]
You are right version 0.9.0 of GLSP is only compatible with sprotty version 0.10.0. The reason for this is that sprotty introduced a complete overhaul of their action definitions with 0.11.0 (switching from a class based approach to an interface+namespace based approach) and deprecated the old way of defining actions.
In the meantime  we have adapted this approach for GLSP and the nightly version of GLSP is compatible with the current sprotty nightly (potentially also 0.11.1 but that has not been tested).

If you want you could try to upgrade to the current nightly/next versions of both projects to see if you issue is resolved:
https://www.npmjs.com/package/@eclipse-glsp/client/v/0.10.0-next.385713d.170
https://www.npmjs.com/package/@eclipse-glsp/theia-integration/v/0.10.0-next.a2c1ac1.122
https://www.npmjs.com/package/sprotty/v/0.11.0-next.e898c14.30

Keep in mind that you probably need some adaptions to your code to resolve compilation errors. This will mostly affect action creation and type checking:
```ts
// Creating a new action
//pre. 0.11.0
new SelectionAction(["elementToSelect"],[])   

//new approach
SelectAction.create({selectedElementIDs: ["elementToSelect"]})  

// Typchecking
//pre 0.11.0
if (isSelectAction(actionToTest)){      
...
}

// new approach
if (SelectAction.is(actionToTest)){      
...
}


```

**Reply by rsoika** [reply]
I have upgraded my project to the nightly build. And it has actually improved a lot! Looks great!

There was only a minor problem: I was unable to resolve my theia  server-Contribution. It looked before like this:

```javascript
import { getPort } from '@eclipse-glsp/protocol';
import { JavaSocketServerContribution, JavaSocketServerLaunchOptions } from '@eclipse-glsp/theia-integration/lib/node';
import { injectable } from '@theia/core/shared/inversify';
import { join, resolve } from 'path';
import { BPMNLanguage } from '../common/bpmn-language';

export const DEFAULT_PORT = 5007;
export const PORT_ARG_KEY = 'GLSP_PORT';
export const LOG_DIR = join(__dirname, '..', '..', 'logs');
const JAR_FILE = resolve(
    join(__dirname, '..', '..', '..', '..', 'open-bpmn.glsp-server', 'target', 'open-bpmn.server-0.0.9-SNAPSHOT-glsp.jar')
);

@injectable()
export class BPMNGLSPServerContribution extends JavaSocketServerContribution {
    readonly id = BPMNLanguage.contributionId;

    createLaunchOptions(): Partial<JavaSocketServerLaunchOptions> {
        return {
            jarPath: JAR_FILE,
            additionalArgs: ['--consoleLog', 'false', '--fileLog', 'true', '--logDir', LOG_DIR],
            socketConnectionOptions: {
                port: getPort(PORT_ARG_KEY, DEFAULT_PORT)
            }
        };
    }
}
```

But the 'getPort()' method seems to no longer exits in @eclipse-glsp/protocol.
I have now hard coded the port
[
            socketConnectionOptions: {
                port: 5007
            }](url)

What would be the recommended implementation now?

**Reply by tortmayr** [reply]
Hi @rsoika ,
Great news! Nice to hear that you managed to upgrade without major issues.
The `getPort()` method has been removed from the common protocol because it was using node based API internally (process) which made it not working in the browser context.
The method has been moved to the theia-integration package and can be imported with:

`import {getPort} from '@eclipse-glsp/theia-integration/lib/node'`

**Reply by rsoika** [reply]
Now afert I have cleaned everything and build up from scratch, I run now into a strange error when starting :

```
index.js:61 TypeError: Cannot read property 'KIND' of undefined
    at vendors-node_modules_eclipse-glsp_client_lib_index_js-node_modules_eclipse-glsp_theia-integra-84627f.bundle.js:15579
    at Object.../node_modules/@eclipse-glsp/theia-integration/lib/browser/diagram/glsp-layout-commands.js (vendors-node_modules_eclipse-glsp_client_lib_index_js-node_modules_eclipse-glsp_theia-integra-84627f.bundle.js:15645)
    at __webpack_require__ (bundle.js:144668)
    at Object.../node_modules/@eclipse-glsp/theia-integration/lib/browser/diagram/index.js (vendors-node_modules_eclipse-glsp_client_lib_index_js-node_modules_eclipse-glsp_theia-integra-84627f.bundle.js:15991)
    at __webpack_require__ (bundle.js:144668)
    at Object.../node_modules/@eclipse-glsp/theia-integration/lib/browser/index.js (vendors-node_modules_eclipse-glsp_client_lib_index_js-node_modules_eclipse-glsp_theia-integra-84627f.bundle.js:16386)
    at __webpack_require__ (bundle.js:144668)
    at Object.../open-bpmn-theia/lib/browser/frontend-module.js (open-bpmn-theia_lib_browser_frontend-module_js-data_image_svg_xml_charset_utf-8_3Csvg_20xmlns-fdcebc.bundle.js:1632)
    at Function.__webpack_require__ (bundle.js:144668)
```

Could it be that I need to upgrade another dependency. For example I have dependencies like:

    "@theia/core": "1.20.0",

I'm just wondering how I was able to successfully build and start with yarn before....

**Reply by tortmayr** [reply]
Hm, what versions are you currently using for `@eclipse-glsp/client` and `@eclipse-glsp/theia-integration`?
Are you using a concrete version or just do you just use the `next` tag?

**Reply by rsoika** [reply]
I do currently using

```
  "devDependencies": {
    "@eclipse-glsp/config": "0.10.0",
    "lerna": "^4.0.0",
    "rimraf": "^2.6.1"
  },

  "dependencies": {
    "@eclipse-glsp/theia-integration": "0.10.0-next.a2c1ac1.122",
    "@eclipse-glsp/client": "0.10.0-next.385713d.170",
    "sprotty": "0.11.0-next.e898c14.30"    
    "@jsonforms/core": "3.0.0-beta.0",
    "@jsonforms/react": "3.0.0-beta.0",
    "@jsonforms/vanilla-renderers": "3.0.0-beta.0",
    "balloon-css": "^0.5.0"
  },
```


and for my -app module for launching the theia app I have still the following dependencies:

```
  "dependencies": {
    "@theia/core": "1.20.0",
    "@theia/editor": "1.20.0",
    "@theia/filesystem": "1.20.0",
    "@theia/markers": "1.20.0",
    "@theia/messages": "1.20.0",
    "@theia/monaco": "1.20.0",
    "@theia/navigator": "1.20.0",
    "@theia/preferences": "1.20.0",
    "@theia/process": "1.20.0",
    "@theia/terminal": "1.20.0",
    "@theia/workspace": "1.20.0",
    "@open-bpmn/open-bpmn-theia": "0.0.9"
  },
```

**Reply by rsoika** [reply]
Maybe my [open-bpmn-theia module](https://github.com/imixs/open-bpmn/tree/master/open-bpmn.glsp-client/open-bpmn-theia) setup is no longer correct? I am still using the code form the workflow-example 0.9.0 

Do you have an updated example working with the nightly build of eclipse-glsp? Then I can verify my own setup.

**Reply by rsoika** [reply]
@tortmayr Maybe this is total nonsenses, but when I try to understand the error message, my browser complains about

```
TypeError: Cannot read property 'KIND' of undefined
    at glsp-layout-commands.ts:40
....
```


in line 40 of glsp-layout-commands.ts I have:

```javascript
export namespace GLSPLayoutCommands {
    export const RESIZE_WIDTH_MIN = 'glsp:' + ResizeElementsCommand.KIND + ':width:min';
```


But seachring for this "ResizeElementsCommand" I can only find a kind of definition (I do not understand what this is) in the file 'layout-elements-action.spec.ts'

```javascript
describe('ResizeElementsCommand', () => {
    let handler: ResizeElementsActionHandler;
    const setModel = (newModel: SModelRoot): void => {
.....
```

And in this function/object there is no KIND property defined. Can this be the problem?

When I look into the code base of glsp-client 0.8.0 I can find the object in 'layout-commands.ts' with a KIND property

```javascript
@injectable()
export class ResizeElementsCommand extends LayoutElementsCommand {
    static readonly KIND = 'layout:resize';
```

Could it be that `ResizeElementsCommand`' is really missing? ....or should I better keep my mouth shut with my JavaScript half-knowledge ;-)

**Comment by tortmayr** [comment]
Hi @rsoika,
I had a quick look at your project at it seems like the problem is that somehow two different versions of `@eclipse-glsp/client` got resolved in your project. The `ResizeElementsCommand` got renamed in between those two versions that's why it cannot be found.
To resolve this issue I recommend to add the glspGLSPversions to  the resolutions block in your root `package.json`:

```json
  "resolutions": {
    "**/@eclipse-glsp/theia-integration": "0.10.0-next.a2c1ac1.122",
    "**/@eclipse-glsp/client": "0.10.0-next.385713d.170",
    "**/@theia/core": "1.20.0",
    "**/@theia/editor": "1.20.0",
    "**/@theia/filesystem": "1.20.0",
    "**/@theia/markers": "1.20.0",
    "**/@theia/messages": "1.20.0",
    "**/@theia/monaco": "1.20.0",
    "**/@theia/navigator": "1.20.0",
    "**/@theia/outline-view": "1.20.0",
    "**/@theia/preferences": "1.20.0",
    "**/@theia/process": "1.20.0",
    "**/@theia/terminal": "1.20.0",
    "**/@theia/variable-resolver": "1.20.0",
    "**/@theia/userstorage": "1.20.0",
    "**/@theia/workspace": "1.20.0",
    "**/@theia/cli": "1.20.0",
    "**/inversify": "5.1.1"
  }
  ```
 This ensures that exactly one version gets resolved during building.
  
 FYI : The `layout-elements-action.spec.ts` is a test file and contains unit tests for the code located in `layout-elements-action.ts`.
  
 I also saw that you are using a `full-wipe.sh` script which deletes the yarn.lock file. I would consider that bad practice. Many dependencies in npm packages have a fluent version range. The yarn.lock file is used to pin this fluent ranges to a concrete version during the initial build. This ensures that subsequent builds are always using with the same dependency tree. There should hardly ever be a reason to complete remove the `yarn.lock` file.
To me it looks like your script is just removing `gitignored` files. You could do the same with the `git clean -xdf` command.

**Reply by rsoika** [reply]
Thanks! Yes adding to lines into the resolutions section solved the problem. Now I can build and run again. Thanks a lot!!

I also improved my project setup according to your recommendations. Thanks.
